Installation failed on Windows 10 (add Win support)

I installed rust stable.

Then I ran the following.

```sh
mkdir webrender-example
cd webrender-example
npm init -y
npm install --save node-webrender
```

Here's what happened:

```
> node-webrender@0.1.8 install C:\code\webrender-example\node_modules\node-webrender
> neon build

neon info forcing rebuild for new build settings
neon info running cargo
    Updating crates.io index
    Updating git repository `https://github.com/servo/webrender.git`
  Downloaded serde_derive v1.0.80
  Downloaded font-loader v0.7.0
  Downloaded lazy_static v1.1.0
  Downloaded time v0.1.40
  Downloaded rayon v1.0.2
  Downloaded neon-build v0.2.0
  Downloaded euclid v0.19.4
.......................Removed for brevity......................
   Compiling winapi-util v0.1.1
   Compiling time v0.1.40
   Compiling atty v0.2.11
   Compiling winit v0.17.2
   Compiling crossbeam-deque v0.2.0
   Compiling sha2 v0.8.0
   Compiling wincolor v1.0.1
   Compiling serde_derive v1.0.80
   Compiling neon-runtime v0.2.0
   Compiling font-loader v0.7.0
   Compiling termcolor v1.0.4
   Compiling env_logger v0.6.0
error: failed to run custom build command for `neon-runtime v0.2.0`
process didn't exit successfully: `C:\code\webrender-example\node_modules\node-webrender\native\target\debug\build\neon-runtime-5c8f3d65c2af5c1e\build-script-build` (exit code: 101)
--- stdout
'Skipping node-gyp installation as part of npm install.'
added 96 packages from 70 contributors and audited 184 packages in 7.067s
found 0 vulnerabilities

cargo:node_arch=x64

--- stderr
thread 'main' panicked at 'Couldn't find node_root_dir in node-gyp output.', libcore\option.rs:1008:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
neon ERR! cargo build failed

Error: cargo build failed
    at Target.<anonymous> (C:\code\webrender-example\node_modules\neon-cli\lib\target.js:121:35)
    at step (C:\code\webrender-example\node_modules\neon-cli\lib\target.js:32:23)
    at Object.next (C:\code\webrender-example\node_modules\neon-cli\lib\target.js:13:53)
    at fulfilled (C:\code\webrender-example\node_modules\neon-cli\lib\target.js:4:58)
    at process._tickCallback (internal/process/next_tick.js:68:7)
npm WARN webrender-example@1.0.0 No description
npm WARN webrender-example@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-webrender@0.1.8 install: `neon build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-webrender@0.1.8 install script.
```

I'm guessing I need more dependencies like the ones listed in debian `g++ cmake pkg-config python libfreetype6 libfreetype6-dev expat libexpat-dev` but I'm not sure how to get all these on Windows.

--- a/native/Cargo.toml
+++ b/native/Cargo.toml
@@ -27,4 +27,4 @@ euclid = "0.19.4"
 log = "0.4.6"
 env_logger = "0.6.0"
 yoga = { version = "0.3.1", features = ["serde_support"] }
-napi-rs = { git = "https://github.com/cztomsik/napi-rs.git", rev = "b6e1c3a" }
+napi-rs = { git = "https://github.com/cztomsik/napi-rs.git", rev = "22596ae" }

--- a/native/build.rs
+++ b/native/build.rs
@@ -1,7 +1,6 @@
-extern crate neon_build;
-
 fn main() {
-    neon_build::setup(); // must be called in build.rs
-
-    // add project-specific build logic here...
+    if cfg!(windows) {
+        println!("cargo:rustc-link-search={}", "C:\\Program Files\\nodejs");
+        println!("cargo:rustc-link-lib={}", "node");
+    }
 }

--- a/native/src/lib.rs
+++ b/native/src/lib.rs
@@ -8,7 +8,7 @@ use std::cell::RefCell;
 use std::io::Write;
 use std::os::unix::net::UnixStream;
 use std::rc::Rc;
-use yoga::{FlexStyle, Size, MeasureMode};
+use yoga::{FlexStyle, MeasureMode, Size};
 
 mod rendering;
 mod resources;
@@ -17,16 +17,23 @@ mod window;
 
 use crate::resources::{OpResource, ResourceManager};
 use crate::surface::{Measure, Surface};
-use crate::window::{EventSender, Window, WindowEvent};
+use crate::window::{Application, EventSender, Window, WindowEvent};
 
 register_module!(node_webrender, init);
+
+thread_local! {
+    static APP: RefCell<Application> = RefCell::new(Application::new());
+}
 
 fn init<'env>(mut ctx: ModuleInitContext) -> Result<Option<Value<'env, Object>>> {
     env_logger::init();
 
+    ctx.export("app_loop_a_bit", callback!(app_loop_a_bit));
     ctx.export("window_create", callback!(window_create));
-    ctx.export("window_handle_events", callback!(window_handle_events));
-    ctx.export("window_get_glyph_indices_and_advances", callback!(window_get_glyph_indices_and_advances));
+    ctx.export(
+        "window_get_glyph_indices_and_advances",
+        callback!(window_get_glyph_indices_and_advances),
+    );
     ctx.export("window_render_surface", callback!(window_render_surface));
     ctx.export("surface_create", callback!(surface_create));
     ctx.export("surface_update", callback!(surface_update));
@@ -49,25 +56,33 @@ fn window_create(ctx: CallContext) -> AnyResult {
     let width = ctx.args[1].f64();
     let height = ctx.args[2].f64();
     let socket_path = ctx.args[3].to_string();
-    let window = Window::new(title, width, height, get_event_sender(&socket_path));
+
+    let window = APP.with(|app| {
+        app.borrow_mut()
+            .create_window(title, width, height, get_event_sender(&socket_path))
+    });
+
     let mut wrapper = ctx.env.create_object();
     ctx.env.wrap(&mut wrapper, window)?;
 
     wrapper.into_result()
 }
 
-fn window_handle_events(ctx: CallContext) -> AnyResult {
-    let window: &mut Window = ctx.args[0].unwrap(ctx.env);
-
-    window.handle_events().into_result()
+fn app_loop_a_bit(_ctx: CallContext) -> AnyResult {
+    APP.with(|app| {
+        app.borrow_mut().loop_a_bit();
+    })
+    .into_result()
 }
 
 fn window_get_glyph_indices_and_advances(ctx: CallContext) -> AnyResult {
-    let window: &mut Window = ctx.args[0].unwrap(ctx.env);
+    let window: &mut Rc<RefCell<Window>> = ctx.args[0].unwrap(ctx.env);
     let font_size = ctx.args[1].i32() as u32;
     let str = ctx.args[2].to_string();
 
-    let (glyph_indices, advances) = window.get_glyph_indices_and_advances(font_size, &str);
+    let (glyph_indices, advances) = window
+        .borrow()
+        .get_glyph_indices_and_advances(font_size, &str);
 
     let mut res_arr = ctx.env.create_array_with_length(2);
     let mut indices_arr = ctx.env.create_array_with_length(glyph_indices.len());
@@ -89,11 +104,10 @@ fn window_get_glyph_indices_and_advances(ctx: CallContext) -> AnyResult {
     res_arr.into_result()
 }
 
-
 fn window_render_surface(ctx: CallContext) -> AnyResult {
     unsafe { RENDER_ENV = Some(ctx.env.borrow_forever()) };
 
-    let window: &mut Window = ctx.args[0].unwrap(ctx.env);
+    let window: &mut Rc<RefCell<Window>> = ctx.args[0].unwrap(ctx.env);
     let surface: &mut Rc<RefCell<Surface>> = ctx.args[1].unwrap(ctx.env);
     let available_width = ctx.args[2].f32();
     let available_height = ctx.args[3].f32();
@@ -103,7 +117,9 @@ fn window_render_surface(ctx: CallContext) -> AnyResult {
         .calculate_layout(available_width, available_height);
 
     ResourceManager::with(|rm| {
-        window.render(&rm.render_ops, &(surface.borrow()));
+        window
+            .borrow_mut()
+            .render(&rm.render_ops, &(surface.borrow()));
     });
 
     unsafe { RENDER_ENV = None };

--- a/native/src/lib.rs
+++ b/native/src/lib.rs
@@ -5,8 +5,6 @@ extern crate serde_derive;
 
 use napi_rs::*;
 use std::cell::RefCell;
-use std::io::Write;
-use std::os::unix::net::UnixStream;
 use std::rc::Rc;
 use yoga::{FlexStyle, MeasureMode, Size};
 
@@ -17,7 +15,7 @@ mod window;
 
 use crate::resources::{OpResource, ResourceManager};
 use crate::surface::{Measure, Surface};
-use crate::window::{Application, EventSender, Window, WindowEvent};
+use crate::window::{Application, EventHandler, Window, WindowEvent};
 
 register_module!(node_webrender, init);
 
@@ -55,11 +53,11 @@ fn window_create(ctx: CallContext) -> AnyResult {
     let title = ctx.args[0].to_string();
     let width = ctx.args[1].f64();
     let height = ctx.args[2].f64();
-    let socket_path = ctx.args[3].to_string();
+    let event_handler = ctx.args[3].cb(ctx.env);
 
     let window = APP.with(|app| {
         app.borrow_mut()
-            .create_window(title, width, height, get_event_sender(&socket_path))
+            .create_window(title, width, height, Box::new(event_handler))
     });
 
     let mut wrapper = ctx.env.create_object();
@@ -68,7 +66,9 @@ fn window_create(ctx: CallContext) -> AnyResult {
     wrapper.into_result()
 }
 
-fn app_loop_a_bit(_ctx: CallContext) -> AnyResult {
+fn app_loop_a_bit(ctx: CallContext) -> AnyResult {
+    unsafe { CURRENT_ENV = Some(ctx.env.borrow_forever()) };
+
     APP.with(|app| {
         app.borrow_mut().loop_a_bit();
     })
@@ -105,7 +105,7 @@ fn window_get_glyph_indices_and_advances(ctx: CallContext) -> AnyResult {
 }
 
 fn window_render_surface(ctx: CallContext) -> AnyResult {
-    unsafe { RENDER_ENV = Some(ctx.env.borrow_forever()) };
+    unsafe { CURRENT_ENV = Some(ctx.env.borrow_forever()) };
 
     let window: &mut Rc<RefCell<Window>> = ctx.args[0].unwrap(ctx.env);
     let surface: &mut Rc<RefCell<Surface>> = ctx.args[1].unwrap(ctx.env);
@@ -121,8 +121,6 @@ fn window_render_surface(ctx: CallContext) -> AnyResult {
             .borrow_mut()
             .render(&rm.render_ops, &(surface.borrow()));
     });
-
-    unsafe { RENDER_ENV = None };
 
     Ok(None)
 }
@@ -209,7 +207,7 @@ fn op_resource_create(ctx: CallContext) -> AnyResult {
 
 fn flex_style_create(ctx: CallContext) -> AnyResult {
     let data = ctx.args[0].to_string();
-    let styles: Rc<Vec<FlexStyle>> = Rc::new(serde_json::from_str(&data).unwrap());
+    let styles: Rc<Vec<FlexStyle>> = Rc::new(serde_json::from_str(&data).expect("invalid style"));
     let mut wrapper = ctx.env.create_object();
 
     debug!("style {:?} -> {:?}", &data, &styles);
@@ -219,30 +217,19 @@ fn flex_style_create(ctx: CallContext) -> AnyResult {
     wrapper.into_result()
 }
 
-struct SocketEventSender {
-    socket: UnixStream,
-}
+impl EventHandler for Ref<Function> {
+    fn handle_event(&mut self, e: WindowEvent) {
+        let env = get_env();
+        let f = env.get_reference_value(self);
+        let payload = env.create_string(&(serde_json::to_string(&e).expect("could not serialize"))).try_into().unwrap();
 
-// TODO: use napi_threadsafe_callback()
-impl EventSender for SocketEventSender {
-    fn send(&mut self, event: WindowEvent) {
-        debug!("sending {:?}", event);
-        let buf: [u8; 12] = unsafe { std::mem::transmute(event) };
-
-        self.socket.write_all(&buf).unwrap();
+        let _ = f.call(None, &[payload]).unwrap();
     }
-}
-
-fn get_event_sender(socket_path: &str) -> Box<EventSender> {
-    let socket_path = std::path::Path::new(socket_path);
-    let socket = UnixStream::connect(&socket_path).unwrap();
-
-    Box::new(SocketEventSender { socket })
 }
 
 impl Measure for Ref<Function> {
     fn measure(&self, w: f32, wm: MeasureMode, h: f32, hm: MeasureMode) -> Size {
-        let env = unsafe { RENDER_ENV.unwrap() };
+        let env = get_env();
         let f = env.get_reference_value(self);
 
         let w = env.create_double(w.into());
@@ -264,7 +251,11 @@ impl Measure for Ref<Function> {
     }
 }
 
-static mut RENDER_ENV: Option<&Env> = None;
+static mut CURRENT_ENV: Option<&Env> = None;
+
+fn get_env<'a>() -> &'a Env {
+    unsafe { CURRENT_ENV.expect("no env available") }
+}
 
 // ---
 // utils bellow
@@ -306,7 +297,7 @@ impl<'env> Helper<'env> for Value<'env, Any> {
     }
 
     fn f64(&self) -> f64 {
-        self.coerce_to_number().unwrap().into()
+        self.coerce_to_number().expect("not a number").into()
     }
 
     fn f32(&self) -> f32 {
@@ -314,7 +305,7 @@ impl<'env> Helper<'env> for Value<'env, Any> {
     }
 
     fn i64(&self) -> i64 {
-        self.coerce_to_number().unwrap().into()
+        self.coerce_to_number().expect("not a number").into()
     }
 
     fn i32(&self) -> i32 {

--- a/native/src/window.rs
+++ b/native/src/window.rs
@@ -25,11 +25,69 @@ use webrender::api::{
 use webrender::Renderer;
 use yoga::Layout;
 
+pub struct Application {
+    events_loop: EventsLoop,
+    windows: Vec<Rc<RefCell<Window>>>,
+}
+
+impl Application {
+    pub fn new() -> Self {
+        let el = EventsLoop::new();
+        let proxy = el.create_proxy();
+        let duration = std::time::Duration::from_millis(200);
+
+        std::thread::spawn(move || loop {
+            std::thread::sleep(duration);
+            let _ = proxy.wakeup();
+        });
+
+        Application {
+            events_loop: el,
+            windows: vec![],
+        }
+    }
+
+    pub fn create_window(
+        &mut self,
+        title: String,
+        w: f64,
+        h: f64,
+        event_sender: Box<EventSender>,
+    ) -> Rc<RefCell<Window>> {
+        let w = Window::new(title, w, h, &self.events_loop, event_sender);
+        let c = Rc::new(RefCell::new(w));
+
+        self.windows.push(c.clone());
+
+        c
+    }
+
+    pub fn loop_a_bit(&mut self) {
+        let windows = &mut self.windows;
+
+        self.events_loop.run_forever(move |e| match e {
+            glutin::Event::Awakened => glutin::ControlFlow::Break,
+            glutin::Event::WindowEvent { window_id, event } => {
+                let w = windows
+                    .iter_mut()
+                    .find(|ref mut w| w.borrow().id == window_id)
+                    .expect("got event from unknown window");
+
+                w.borrow_mut().handle_event(&event);
+
+                glutin::ControlFlow::Continue
+            }
+            _ => glutin::ControlFlow::Continue,
+        });
+    }
+}
+
 // TODO: split
+// TODO: drop (from app)
 pub struct Window {
     // win
+    id: glutin::WindowId,
     gl_window: GlWindow,
-    events_loop: Rc<RefCell<EventsLoop>>,
     dpi: f64,
 
     // rendering
@@ -51,8 +109,14 @@ pub struct Window {
 }
 
 impl Window {
-    pub fn new(title: String, width: f64, height: f64, event_sender: Box<EventSender>) -> Self {
-        let (gl_window, events_loop) = Window::create_gl_window(title, width, height);
+    pub fn new(
+        title: String,
+        width: f64,
+        height: f64,
+        events_loop: &EventsLoop,
+        event_sender: Box<EventSender>,
+    ) -> Self {
+        let gl_window = Window::create_gl_window(title, width, height, events_loop);
 
         let dpi = gl_window.get_hidpi_factor();
         let (tx, rx) = channel();
@@ -61,8 +125,8 @@ impl Window {
         let (font, font_index, font_key) = Window::load_font(&mut api);
 
         let mut w = Window {
+            id: gl_window.id(),
             gl_window,
-            events_loop: Rc::new(RefCell::new(events_loop)),
             dpi,
 
             api,
@@ -85,56 +149,36 @@ impl Window {
         w
     }
 
-    pub fn handle_events(&mut self) {
+    pub fn handle_event(&mut self, event: &glutin::WindowEvent) {
         let mut should_redraw = false;
 
-        // this is easier than rethinking/rewriting everything just because of unique access needed in closure
-        {
-            let evl_rc = self.events_loop.clone();
-            let mut events_loop = evl_rc.borrow_mut();
+        match *event {
+            glutin::WindowEvent::HiDpiFactorChanged(dpi) => self.dpi = dpi,
 
-            events_loop.poll_events(|glutin_event| {
-                match glutin_event {
-                    glutin::Event::WindowEvent { event, .. } => {
-                        debug!("Event {:?}", event);
+            glutin::WindowEvent::CloseRequested => self.handle_close(),
 
-                        match event {
-                            glutin::WindowEvent::HiDpiFactorChanged(dpi) => self.dpi = dpi,
+            glutin::WindowEvent::Resized(size) => {
+                self.handle_resize(size);
+                should_redraw = true
+            }
 
-                            glutin::WindowEvent::CloseRequested => self.handle_close(),
+            glutin::WindowEvent::CursorMoved { position, .. } => self.mouse_position = position,
 
-                            glutin::WindowEvent::Resized(size) => {
-                                self.handle_resize(size);
-                                should_redraw = true
-                            }
+            glutin::WindowEvent::MouseInput { state, .. } => self.handle_mouse(state),
 
-                            glutin::WindowEvent::CursorMoved { position, .. } => {
-                                self.mouse_position = position
-                            }
+            // TODO: scroll x
+            glutin::WindowEvent::MouseWheel { delta, .. } => {
+                let y = match delta {
+                    glutin::MouseScrollDelta::PixelDelta(point) => (point.y as f32),
+                    glutin::MouseScrollDelta::LineDelta(_, dy) => 30. * dy,
+                };
 
-                            glutin::WindowEvent::MouseInput { state, .. } => {
-                                self.handle_mouse(state)
-                            }
+                self.scroll(y);
+            }
 
-                            // TODO: scroll x
-                            glutin::WindowEvent::MouseWheel { delta, .. } => {
-                                let y = match delta {
-                                    glutin::MouseScrollDelta::PixelDelta(point) => (point.y as f32),
-                                    glutin::MouseScrollDelta::LineDelta(_, dy) => 30. * dy,
-                                };
+            glutin::WindowEvent::ReceivedCharacter(ch) => self.handle_char(ch),
 
-                                self.scroll(y);
-                            }
-
-                            glutin::WindowEvent::ReceivedCharacter(ch) => self.handle_char(ch),
-
-                            _ => {}
-                        }
-                    }
-
-                    _ => {}
-                }
-            });
+            _ => {}
         }
 
         if should_redraw {
@@ -283,7 +327,12 @@ impl Window {
         (glyph_indices, advances)
     }
 
-    fn create_gl_window(title: String, width: f64, height: f64) -> (GlWindow, EventsLoop) {
+    fn create_gl_window(
+        title: String,
+        width: f64,
+        height: f64,
+        events_loop: &EventsLoop,
+    ) -> GlWindow {
         let window_builder = glutin::WindowBuilder::new()
             .with_title(title)
             .with_dimensions(glutin::dpi::LogicalSize::new(width, height));
@@ -292,13 +341,12 @@ impl Window {
                 opengl_version: (3, 2),
                 opengles_version: (3, 0),
             });
-        let events_loop = EventsLoop::new();
-        let gl_window = GlWindow::new(window_builder, context_builder, &events_loop).unwrap();
+        let gl_window = GlWindow::new(window_builder, context_builder, events_loop).unwrap();
 
         unsafe { gl_window.make_current().ok() };
         gl_window.show();
 
-        (gl_window, events_loop)
+        gl_window
     }
 
     fn create_api(

--- a/native/src/window.rs
+++ b/native/src/window.rs
@@ -24,6 +24,7 @@ use webrender::api::{
 };
 use webrender::Renderer;
 use yoga::Layout;
+use serde::Serialize;
 
 pub struct Application {
     events_loop: EventsLoop,
@@ -34,9 +35,9 @@ impl Application {
     pub fn new() -> Self {
         let el = EventsLoop::new();
         let proxy = el.create_proxy();
-        let duration = std::time::Duration::from_millis(200);
+        let duration = std::time::Duration::from_millis(30);
 
-        std::thread::spawn(move || loop {
+        std::thread::spawn(move ||loop {
             std::thread::sleep(duration);
             let _ = proxy.wakeup();
         });
@@ -52,9 +53,9 @@ impl Application {
         title: String,
         w: f64,
         h: f64,
-        event_sender: Box<EventSender>,
+        event_handler: Box<EventHandler>,
     ) -> Rc<RefCell<Window>> {
-        let w = Window::new(title, w, h, &self.events_loop, event_sender);
+        let w = Window::new(title, w, h, &self.events_loop, event_handler);
         let c = Rc::new(RefCell::new(w));
 
         self.windows.push(c.clone());
@@ -105,7 +106,7 @@ pub struct Window {
 
     // events
     mouse_position: LogicalPosition,
-    event_sender: Box<EventSender>,
+    event_handler: Box<EventHandler>,
 }
 
 impl Window {
@@ -114,7 +115,7 @@ impl Window {
         width: f64,
         height: f64,
         events_loop: &EventsLoop,
-        event_sender: Box<EventSender>,
+        event_handler: Box<EventHandler>,
     ) -> Self {
         let gl_window = Window::create_gl_window(title, width, height, events_loop);
 
@@ -141,7 +142,7 @@ impl Window {
             font_key,
 
             mouse_position: LogicalPosition::new(0., 0.),
-            event_sender,
+            event_handler,
         };
 
         w.send_initial_frame();
@@ -215,7 +216,7 @@ impl Window {
     }
 
     fn handle_close(&mut self) {
-        self.event_sender.send(WindowEvent::Close);
+        self.event_handler.handle_event(WindowEvent::Close);
     }
 
     fn handle_resize(&mut self, size: LogicalSize) {
@@ -228,16 +229,16 @@ impl Window {
             .context()
             .resize(PhysicalSize::from_logical(size, self.dpi));
 
-        self.event_sender
-            .send(WindowEvent::Resize(size.width as f32, size.height as f32));
+        self.event_handler
+            .handle_event(WindowEvent::Resize(size.width as f32, size.height as f32));
     }
 
     fn handle_char(&mut self, ch: char) {
-        self.event_sender.send(WindowEvent::KeyPress(ch))
+        self.event_handler.handle_event(WindowEvent::KeyPress(ch))
     }
 
     fn handle_mouse(&mut self, state: glutin::ElementState) {
-        /*self.event_sender.send(WindowEvent::MouseInput(
+        /*self.event_handler.send(WindowEvent::MouseInput(
             self.mouse_position.x,
             self.mouse_position.y,
         ))*/
@@ -247,7 +248,7 @@ impl Window {
                 let res = self.hit_test(cursor);
 
                 for it in res.items {
-                    self.event_sender.send(WindowEvent::Click(it.tag.0 as u32))
+                    self.event_handler.handle_event(WindowEvent::Click(it.tag.0 as u32))
                 }
             }
 
@@ -475,7 +476,7 @@ impl RenderNotifier for Notifier {
 
 pub struct Msg {}
 
-#[derive(Debug)]
+#[derive(Debug, Serialize)]
 pub enum WindowEvent {
     Close,
 
@@ -491,6 +492,6 @@ pub enum WindowEvent {
     Click(u32),
 }
 
-pub trait EventSender {
-    fn send(&mut self, event: WindowEvent);
+pub trait EventHandler {
+    fn handle_event(&mut self, event: WindowEvent);
 }

--- a/package.json
+++ b/package.json
@@ -21,7 +21,7 @@
     "native/src"
   ],
   "scripts": {
-    "install": "neon build",
+    "install": "neon build --release",
     "prepublishOnly": "neon build && tsc -d",
     "example": "RUST_LOG=debug ts-node -T example/example.tsx",
     "example-calc": "RUST_LOG=debug ts-node -T example/react-calculator.tsx"

--- a/src/core/TextContainer.ts
+++ b/src/core/TextContainer.ts
@@ -14,6 +14,8 @@ import { NativeApi, N } from './nativeApi'
 const native: NativeApi = require('../../native')
 
 // TODO: this should be (at least partially) in rust
+// note that we expect updateGlyphs/Brush to be called only from measure func
+// (because of double borrow & and also because of skipping unnecessary updates)
 export class TextContainer {
   ref
   children = []
@@ -71,8 +73,6 @@ export class TextContainer {
     this.fontInstanceKey = [1, fontSize]
     this.color = color
     this.lineHeight = lineHeight
-
-    this.updateBrush()
   }
 
   updateContent() {

--- a/src/core/Window.ts
+++ b/src/core/Window.ts
@@ -33,9 +33,6 @@ class Window implements Container<Surface | TextContainer> {
     this.height = height
 
     this.root.update({ layout: FILL_LAYOUT })
-
-    // needed because there is no proper threading yet
-    setInterval(() => native.window_handle_events(this.ref), 1000 / 30)
   }
 
   // TODO
@@ -134,6 +131,14 @@ const decodeEvent = (b: Buffer) => {
       throw new Error(`unknown event ${type} ${b}`)
   }
 }
+
+const handleEvents = () => {
+  native.app_loop_a_bit()
+
+  setImmediate(handleEvents)
+}
+
+handleEvents()
 
 // TODO: refactor rust so ResourceManager is separate from Window
 // (almost done, we just need to get glyph indices & advances)

--- a/src/core/Window.ts
+++ b/src/core/Window.ts
@@ -1,4 +1,3 @@
-import * as net from 'net'
 import { Container } from './types'
 import { Surface, TextContainer } from '.'
 import { NativeApi, N } from './nativeApi'
@@ -15,13 +14,14 @@ class Window implements Container<Surface | TextContainer> {
   height
   onClose = DEFAULT_CLOSE
   onKeyPress = e => {}
+  renderScheduled = false
 
   constructor(title, width = 800, height = 600) {
     this.ref = native.window_create(
       title,
       width,
       height,
-      readEvents(e => this.handleEvent(e))
+      (evStr => this.handleEvent(JSON.parse(evStr)))
     )
 
     windowCount++
@@ -37,22 +37,31 @@ class Window implements Container<Surface | TextContainer> {
 
   // TODO
   handleEvent(event) {
-    switch (event.type) {
-      case 'Close':
-        return this.onClose()
-      case 'KeyPress':
-        return this.onKeyPress(event)
-      case 'Resize':
-        this.width = event.width
-        this.height = event.height
-        this.render()
-        return
-      case 'Click':
-        __callbacks[event.callbackIndex]()
-        return
-      default:
-        throw new Error('unknown event')
+    if (event === 'Close') {
+      this.onClose()
+      return
     }
+
+    if ('KeyPress' in event) {
+      this.onKeyPress({ ch: event.KeyPress })
+      return
+    }
+
+    if ('Resize' in event) {
+      let [width, height] = event.Resize
+
+      this.width = width
+      this.height = height
+      this.renderLater()
+      return
+    }
+
+    if ('Click' in event) {
+      __callbacks[event.Click]()
+      return
+    }
+
+    throw new Error('unknown event')
   }
 
   appendChild(child) {
@@ -75,6 +84,18 @@ class Window implements Container<Surface | TextContainer> {
       this.height
     )
   }
+
+  renderLater() {
+    if (this.renderScheduled) {
+      return
+    }
+
+    setImmediate(() => {
+      this.render()
+      this.renderScheduled = false
+    })
+    this.renderScheduled = true
+  }
 }
 
 let windowCount = 0
@@ -83,52 +104,6 @@ const DEFAULT_CLOSE = () => {
   // TODO: open/close()?
   if (!--windowCount) {
     process.exit()
-  }
-}
-
-// TODO: optimize
-const readEvents = handler => {
-  const EVENT_SIZE = 12
-  const socketPath = `/tmp/nwr-${Date.now()}`
-
-  const server = net.createServer(socket => {
-    let rest = Buffer.alloc(0)
-
-    socket.on('data', b => {
-      b = Buffer.concat([rest, b])
-
-      let i = 0
-
-      while (i + EVENT_SIZE <= b.length) {
-        handler(decodeEvent(b.slice(i, i + EVENT_SIZE)))
-        i += EVENT_SIZE
-      }
-
-      rest = b.slice(i)
-    })
-
-    console.debug('listening for events')
-  })
-  server.listen(socketPath)
-  return socketPath
-}
-
-const decodeEvent = (b: Buffer) => {
-  const TYPES = ['Close', 'Resize', 'KeyPress', 'Click']
-
-  const type = TYPES[b.readInt8(0)]
-
-  switch (type) {
-    case 'Close':
-      return { type }
-    case 'Resize':
-      return { type, width: b.readFloatLE(4), height: b.readFloatLE(8) }
-    case 'KeyPress':
-      return { type, ch: String.fromCharCode(b.readInt32LE(4)) }
-    case 'Click':
-      return { type, callbackIndex: b.readUInt32LE(4) }
-    default:
-      throw new Error(`unknown event ${type} ${b}`)
   }
 }
 

--- a/src/core/nativeApi.ts
+++ b/src/core/nativeApi.ts
@@ -38,7 +38,7 @@ export interface NativeApi {
     title: string,
     width: number,
     hieght: number,
-    socketPath: string
+    event_handler: (evStr: string) => void
   ) => N.Window
 
   window_render_surface: (

--- a/src/core/nativeApi.ts
+++ b/src/core/nativeApi.ts
@@ -32,14 +32,14 @@ export module N {
 }
 
 export interface NativeApi {
+  app_loop_a_bit: () => void
+
   window_create: (
     title: string,
     width: number,
     hieght: number,
     socketPath: string
   ) => N.Window
-
-  window_handle_events: (win: N.Window) => void
 
   window_render_surface: (
     win: N.Window,

--- a/src/react/reconciler.ts
+++ b/src/react/reconciler.ts
@@ -104,7 +104,7 @@ function insertBefore(parent, child, before) {
 }
 
 function resetAfterCommit(window) {
-  window.render()
+  window.renderLater()
 }
 
 declare global {

